test/common/router/rds_impl_test.cc

```diff
@@ -276,5 +278,80 @@ TEST_F(RdsImplTest, FailureArray) {
   EXPECT_EQ(1UL, store_.counter("foo.rds.update_failure").value());
 }
 
+class RouteConfigProviderManagerImplTest : public testing::Test {
+public:
+  RouteConfigProviderManagerImplTest() {}
+  ~RouteConfigProviderManagerImplTest() {}
+
+  NiceMock<Runtime::MockLoader> runtime_;
+  NiceMock<Upstream::MockClusterManager> cm_;
+  Event::MockDispatcher dispatcher_;
+  NiceMock<Runtime::MockRandomGenerator> random_;
+  NiceMock<LocalInfo::MockLocalInfo> local_info_;
+  Stats::IsolatedStoreImpl store_;
+  NiceMock<ThreadLocal::MockInstance> tls_;
+  NiceMock<Init::MockManager> init_manager_;
+  RouteConfigProviderManagerImpl route_config_provider_manager_{runtime_, dispatcher_, random_,
+                                                                local_info_, tls_};
+};
+
+TEST_F(RouteConfigProviderManagerImplTest, Basic) {
+  init_manager_.initialize();
+
+  std::string config_json = R"EOF(
+    {
+      "cluster": "foo_cluster",
+      "route_config_name": "foo_route_config",
+      "refresh_delay_ms": 1000
+    }
+    )EOF";
+
+  Json::ObjectSharedPtr config = Json::Factory::loadFromString(config_json);
+
+  // Get a RouteConfigProvider. This one should create an entry in the RouteConfigProviderManager.
+  RouteConfigProviderSharedPtr provider = route_config_provider_manager_.getRouteConfigProvider(
+      *config, cm_, store_, "foo_prefix", init_manager_);
+  // Because this get has the same cluster and route_config_name, the provider returned is just a
+  // shared_ptr to the same provider as the one above.
+  RouteConfigProviderSharedPtr provider2 = route_config_provider_manager_.getRouteConfigProvider(
+      *config, cm_, store_, "foo_prefix", init_manager_);
+  // So this means that both shared_ptrs should be the same.
+  EXPECT_EQ(provider, provider2);
```

Comment:
Have you considered using the `shared_ptr` `use_count()` method to validate the behaviors around reference counting?